### PR TITLE
Fix GraphQL batch proposals

### LINKED ISSUE #10327: [Bug]: Batch proposals in graphQL not working
### Problem encountered

In the overnight run the test `test_graphQl_CheckMetaData` is failing because a grapqhQL call that list proposals in struggling when a batch proposal exists:
```
            for prop in json_data["data"]["proposalsConnection"]["edges"]:
>               prop_node = prop["node"]
E               TypeError: 'NoneType' object is not subscriptable
```

Running a query through the graphQL playground on the same data gives me:
image

which is likely because in the graphQL schema the `terms` field is non-nullable but I guess is empty when its a batch proposal.

Also a field for `batchTerms` seems to be missing from the graphQL type, but exists in the gRPC endpoint.

### Observed behaviour

GraphQL is not happy querying batch proposals.

### Expected behaviour

It should be possible to query batch proposals, and the batched terms, through graphQL

### Steps to reproduce

```bash
1. Submit `xyz` transaction
2. With this payload
3. Query `abc` API
4. Observe the error
```

### Software version

develop

### Failing test

_No response_

### Jenkins run

_No response_

### Configuration used

_No response_

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by karlem**
Close #10327

### CODE REVIEW

**Comment by karlem** [comment]
Solved by adding union type to the proposal endpoints. It can be queried like so:

```
{
	proposal(id: "id") {
    ... on Proposal{
      reference
    }
    ... on BatchProposal {
      reference,
      subProposals {
        id
        reference
        batchId
      }
    }
  }
  
  proposalsConnection {
    edges {
      node {
        ... on Proposal{
          reference
        }
        ... on BatchProposal {
          id
          reference,
          subProposals {
            id
            reference
            batchId
          }
        }
      }
    }
  }
}
```

File: datanode/gateway/graphql/schema.graphql [hunk 1]
```diff
@@ -5926,10 +5985,12 @@ enum TransferDirection {
   ToOrFrom
 }
 
+union ProposalNode = Proposal | BatchProposal
+
 "Edge type containing the proposals and cursor information returned by a ProposalsConnection"
 type ProposalEdge {
   "The proposal data"
-  node: Proposal!
+  node: ProposalNode!
```
Comment (wwestgarth) [inline datanode/gateway/graphql/schema.graphql:5993]: We should do a proper deprecation here and have both `Proposal` and `ProposalNode` fields, but mark `Proposal` as deprecated.

File: datanode/gateway/graphql/schema.graphql [hunk 2]
```diff
@@ -4873,6 +4872,31 @@ type ProposalTerms {
   change: ProposalChange!
 }
 
+"Terms for a batch governance proposal on Vega"
```
Comment (candida-d) [inline datanode/gateway/graphql/schema.graphql:4875]:
```suggestion
"Terms for a batch governance proposal"
```

File: datanode/gateway/graphql/schema.graphql [hunk 3]
```diff
@@ -4873,6 +4872,31 @@ type ProposalTerms {
...
+type BatchProposalTermsChange {
+  """
+  RFC3339Nano time and date when this proposal is executed (if passed). Note that it has to be after closing date time.
```
Comment (candida-d) [inline datanode/gateway/graphql/schema.graphql:4878]:
```suggestion
  RFC3339Nano time and date when this proposal will be executed, if it passes.
```
I don't think the note is relevant here as this is querying for an existing proposal so there's not much the user can do about it at this stage.

File: datanode/gateway/graphql/schema.graphql [hunk 4]
```diff
@@ -4873,6 +4872,31 @@ type ProposalTerms {
...
+  Constrained by "minEnactInSeconds" and "maxEnactInSeconds" network parameters.
+  Note: Optional as free form proposals do not require it.
+  """
+  enactmentDatetime: Timestamp
+
+  "Actual change being introduced by the proposal - action the proposal triggers if passed and enacted."
+  change: ProposalChange!
+}
+
+"The rationale behind the proposal"
```
Comment (candida-d) [inline datanode/gateway/graphql/schema.graphql:4888]:
```suggestion
"The rationale for the proposal"
```

File: datanode/gateway/graphql/schema.graphql [hunk 5]
```diff
@@ -4960,9 +4984,42 @@ type Proposal {
   requiredLpParticipation: String
 }
 
+type BatchProposal {
+  "Proposal ID that is filled by Vega once proposal reaches the network"
```
Comment (candida-d) [inline datanode/gateway/graphql/schema.graphql:4988]:
```suggestion
  "Proposal ID that is provided by Vega once proposal reaches the network"
```

File: datanode/gateway/graphql/schema.graphql [hunk 6]
```diff
@@ -4960,9 +4984,42 @@ type Proposal {
...
+  id: ID
+  "A UUID reference to aid tracking proposals on Vega"
+  reference: String!
+  "Party that prepared the proposal"
+  party: Party!
+  "State of the proposal"
+  state: ProposalState!
+  "RFC3339Nano time and date when the proposal reached Vega network"
```
Comment (candida-d) [inline datanode/gateway/graphql/schema.graphql:4996]:
```suggestion
  "RFC3339Nano time and date when the proposal reached the network"
```

File: datanode/gateway/graphql/schema.graphql [hunk 7]
```diff
@@ -4960,9 +4984,42 @@ type Proposal {
...
+  datetime: Timestamp!
+  "Rationale behind the proposal"
+  rationale: ProposalRationale!
+  "Votes cast for this proposal"
+  votes: ProposalVotes!
+  "Why the proposal was rejected by the core"
```
Comment (candida-d) [inline datanode/gateway/graphql/schema.graphql:5002]:
```suggestion
  "Reason the proposal was rejected"
```

File: datanode/gateway/graphql/schema.graphql [hunk 8]
```diff
@@ -4960,9 +4984,42 @@ type Proposal {
...
+  rejectionReason: ProposalRejectionReason
+  "Error details of the rejectionReason"
```
Comment (candida-d) [inline datanode/gateway/graphql/schema.graphql:5004]:
```suggestion
  "Details of the rejection reason"
```

File: datanode/gateway/graphql/schema.graphql [hunk 9]
```diff
@@ -4960,9 +4984,42 @@ type Proposal {
...
+  errorDetails: String
+  "Required majority for this proposal to succeed"
+  requiredMajority: String!
+  "Required participation for this proposal to succeed"
+  requiredParticipation: String!
+  "Equity-like share required for a market amendment proposal to be enacted (if not met, the proposal will not be enacted), represented as a fraction that can be converted to a percentage"
```
Comment (candida-d) [inline datanode/gateway/graphql/schema.graphql:5010]:
```suggestion
  "Equity-like share required for a market amendment proposal to be enacted, represented as a fraction that can be converted to a percentage. If not met, the proposal will not be enacted"
```

File: datanode/gateway/graphql/schema.graphql [hunk 10]
```diff
@@ -4960,9 +4984,42 @@ type Proposal {
...
+  requiredLpMajority: String
+  "The market share of LPs' equity-like share that must take part in the market amendment vote for the proposal to pass. This means the votes of LPs that have submitted more liquidity to that market, or have been LPs from the start carry more weight. If it requires 50% of a market's equity-like share for a majority, and the proposal receives only YES votes but only LPs with 49% of the equity-like share voted, the proposal will not pass"
```
Comment (candida-d) [inline datanode/gateway/graphql/schema.graphql:5012]:
```suggestion
  "The market share of LPs' equity-like share that must take part in a market amendment vote for the proposal to pass. This means the votes of LPs that have submitted more liquidity to that market, or have been LPs from the start carry more weight. If it requires 50% of a market's equity-like share for a majority, and the full batch of proposals receives all YES votes but only LPs with 49% of the equity-like share voted, the proposal will not pass"
```

File: datanode/gateway/graphql/schema.graphql [hunk 11]
```diff
@@ -4960,9 +4984,42 @@ type Proposal {
...
+  requiredLpParticipation: String
+  "Child proposals that are part of the batch"
+  subProposals: [ProposalDetail]
+}
+
 type ProposalDetail {
   "Proposal ID that is provided by Vega once proposal reaches the network"
   id: ID
+  "Parent proposal ID that is provided by Vega once proposal reaches the network"
```
Comment (candida-d) [inline datanode/gateway/graphql/schema.graphql:5021]:
```suggestion
  "Batch proposal ID that is provided by Vega once proposal reaches the network"
```

File: datanode/gateway/graphql/schema.graphql [hunk 12]
```diff
@@ -4971,8 +5028,10 @@ type ProposalDetail {
   state: ProposalState!
   "RFC3339Nano time and date when the proposal reached the Vega network"
   datetime: Timestamp!
-  "Terms of the proposal"
-  terms: ProposalTerms!
+  "Terms of the proposal - in case of single proposal"
+  terms: ProposalTerms
+  "Terms of the proposal - in case of batch proposal"
```
Comment (candida-d) [inline datanode/gateway/graphql/schema.graphql:5033]:
```suggestion
  "Terms of the proposal for single proposal"
  terms: ProposalTerms
  "Terms of the proposal for a batch proposal"
```

File: datanode/gateway/graphql/schema.graphql [hunk 13]
```diff
@@ -4971,8 +5028,10 @@ type ProposalDetail {
...
+  batchTerms: BatchProposalTerms
   "Rationale behind the proposal"
   rationale: ProposalRationale!
   "Why the proposal was rejected by the core"
```
Comment (candida-d) [inline datanode/gateway/graphql/schema.graphql:5037]:
```suggestion
  "Why the proposal was rejected"
```

File: datanode/gateway/graphql/schema.graphql [hunk 14]
```diff
@@ -4960,9 +4984,42 @@ type Proposal {
...
+  errorDetails: String
+  "Required majority for this proposal to succeed"
+  requiredMajority: String!
+  "Required participation for this proposal to succeed"
+  requiredParticipation: String!
+  "Equity-like share required for a market amendment proposal to be enacted (if not met, the proposal will not be enacted), represented as a fraction that can be converted to a percentage"
+  requiredLpMajority: String
+  "The market share of LPs' equity-like share that must take part in the market amendment vote for the proposal to pass. This means the votes of LPs that have submitted more liquidity to that market, or have been LPs from the start carry more weight. If it requires 50% of a market's equity-like share for a majority, and the proposal receives only YES votes but only LPs with 49% of the equity-like share voted, the proposal will not pass"
+  requiredLpParticipation: String
+  "Child proposals that are part of the batch"
```
Comment (candida-d) [inline datanode/gateway/graphql/schema.graphql:5014]:
```suggestion
  "Proposals that are part of the batch"
```

File: datanode/gateway/graphql/proposa_edge_resolver.go [hunk 15]
```diff
@@ -0,0 +1,79 @@
+// Copyright (C) 2023 Gobalsky Labs Limited
```
Comment (wwestgarth) [inline datanode/gateway/graphql/proposa_edge_resolver.go:1]: Minor: type in this file name `proposa -> proposal`

**Review by wwestgarth — DISMISSED** [review]

**Review by jeremyletang — APPROVED** [review]

**Pull request merged**